cpp/src/table/device_table.cuh

```diff
@@ -328,38 +402,33 @@ __device__ inline hash_value_type hash_row(device_table const& t,
       hash_value_type{0}, hash_combiner);
 }
 
-namespace {
-struct copy_element {
-  template <typename T>
-  __device__ inline void operator()(gdf_column const& target,
-                                    gdf_size_type target_index,
-                                    gdf_column const& source,
-                                    gdf_size_type source_index) {
-    static_cast<T*>(target.data)[target_index] =
-        static_cast<T const*>(source.data)[source_index];
-  }
-};
-}  // namespace
-
 /**
  * @brief  Copies a row from a source table to a target table.
  *
  * This device function should be called by a single thread and the thread
  * will copy all of the elements in the row from one table to the other.
  *
- * FIXME: Does NOT set null bitmask for the target row.
+ * @note If the `update_target_bitmask` template argument is `true`, then this
+ * function will update the target bitmask (if it exists) to match the bitmask
+ * for the source element. A non-existant source bitmask is assumed to mean the
+ * source element is non-null. However, this operation is very inefficient and
+ * should be done outside of this function when possible.
  *
  * @param[in,out] The table whose row will be updated
  * @param[in] The index of the row to update in the target table
  * @param[in] source The table whose row will be copied
  * @param source_row_index The index of the row to copy in the source table
+ * @tparam update_target_bitmask Flag indicating if the target bitmask should be
+ * updated
  */
+template <bool update_target_bitmask = true>
 __device__ inline void copy_row(device_table const& target,
                                 gdf_size_type target_index,
                                 device_table const& source,
                                 gdf_size_type source_index) {
   for (gdf_size_type i = 0; i < target.num_columns(); ++i) {
-    cudf::type_dispatcher(target.get_column(i)->dtype, copy_element{},
+    cudf::type_dispatcher(target.get_column(i)->dtype,
```

Comment:
should we add some asserts here? this function could be misused if its set to update_target_bitmark, but the target does not have valids allocated

Reply:
In `copy_element` I check for if the `target_mask != nullptr` to avoid that exact pitfall. Doing all those checks is part of what makes this such an inefficient operation and why I caution against using it. That said, sometimes it is necessary and/or nice to have something that "just works" and then worry about optimization later.